static/app/api/apiOptions.spec.ts

```diff
@@ -0,0 +1,175 @@
+      const options = apiOptions('/projects/$orgSlug/', {
+        staleTime: 0,
+        // @ts-expect-error Invalid path parameter
+        path: {orgSlug: 'my-org', invalidParam: 'invalid'},
+      });
+
+      expectTypeOf(options.queryFn).returns.toEqualTypeOf<QueryFunctionResult<never>>();
+    });
+
+    test('should not allow excess path parameters', () => {
+      const options = apiOptions('/projects/$orgSlug/', {
+        staleTime: 0,
+        // @ts-expect-error Excess path parameter
+        path: {orgSlug: 'my-org', extraParam: 'extra'},
+      });
+
+      expectTypeOf(options.queryFn).returns.toEqualTypeOf<QueryFunctionResult<never>>();
+    });
+
+    test('should require path params for paths with parameters', () => {
+      expect(() => {
+        const options = apiOptions('/projects/$orgSlug/', {
+          staleTime: 0,
+          // @ts-expect-error Missing required path parameter
+          path: {},
+        });
+
+        expectTypeOf(options.queryFn).returns.toEqualTypeOf<QueryFunctionResult<never>>();
+      }).toThrow('Missing path param: orgSlug');
+    });
+
+    test('should not allow empty path parameters for paths without parameters', () => {
+      const options = apiOptions('/projects/', {
+        staleTime: 0,
+        // @ts-expect-error Empty path parameters not allowed
+        path: {},
+      });
+
+      expectTypeOf(options.queryFn).returns.toEqualTypeOf<QueryFunctionResult<never>>();
+    });
+
+    test('should not need path params for paths without parameters', () => {
+      const options = apiOptions('/projects/', {
+        staleTime: 0,
+      });
+
+      expectTypeOf(options.queryFn).returns.toEqualTypeOf<QueryFunctionResult<never>>();
+    });
+
+    test('should allow string or number path parameters', () => {
+      const options = apiOptions('/items/$id/', {
+        staleTime: 0,
+        path: {id: 123},
+      });
+
+      expectTypeOf(options.queryFn).returns.toEqualTypeOf<QueryFunctionResult<never>>();
+
+      const options2 = apiOptions('/items/$id/', {
+        staleTime: 0,
+        path: {id: 'abc'},
+      });
+
+      expectTypeOf(options2.queryFn).returns.toEqualTypeOf<QueryFunctionResult<never>>();
+    });
+
+    test('should default to never for unknown API paths', () => {
+      const options = apiOptions('/unknown/$param/', {
+        staleTime: 0,
+        path: {param: 'value'},
+      });
+
+      expectTypeOf(options.queryFn).returns.toEqualTypeOf<QueryFunctionResult<never>>();
+    });
+
+    test('should allow providing manual data type', () => {
+      const options = apiOptions.ReturnType<number>()('/foo/$bar', {
+        staleTime: 0,
+        path: {bar: 'baz'},
+      });
+
+      expectTypeOf(options.queryFn).returns.toEqualTypeOf<QueryFunctionResult<number>>();
+    });
+
+    test('should have a default select that extracts content', () => {
+      const options = apiOptions.ReturnType<number>()('/items/$id/', {
```

Comment:
wouldn't we need to migrate every single call site to this funky name

Reply:
the name is up for debate, it just needs a name because if you do `apiOptions<number>()`, you destroy the other 3 generics that are passed in and then inference for path params won’t work anymore.

as for migration, the idea is that we add paths to `apiDefinition` to get inference, and if we do that, we don’t need to specify a return type, because we can infer it.

Think of it as a contract like: “this api endpoint has these params and returns this type”. I think a central inventory like that is beneficial to have, even if it means moving a lot of types into the `api` directory.